[ValueTracking] Compute KnownFP state from recursive select/phi.

davemgreen opened the pull request:
Given a recursive phi with select:
 %p = phi [ 0, entry ], [ %sel, loop]
 %sel = select %c, %other, %p

The fp state can be calculated using the knowledge that the select/phi pair can only be the initial state (0 here) or from %other. This adds a short-cut into computeKnownFPClass for PHI to detect that the select is recursive back to the phi, and if so use the state from the other operand.

This helps to address a regression from #83200.

pawosm-arm reviewed (approved):
Tested locally, this indeed restores the lost transformation.

davemgreen committed:
[ValueTracking] Compute KnownFP state from recursive select/phi.

Given a recursive phi with select:
 %p = phi [ 0, entry ], [ %sel, loop]
 %sel = select %c, %other, %p

The fp state can be calculated using the knowledge that the select/phi pair can
only be the initial state (0 here) or from %other. This adds a short-cut into
computeKnownFPClass for PHI to detect that the select is recursive back to the
phi, and if so use the state from the other operand.

This helps to address a regression from #83200.

dtcxzyw commented:
I think this approach is also applicable to `computeKnownBits`. I'll post a patch later :)

davemgreen commented:
> I think this approach is also applicable to `computeKnownBits`. I'll post a patch later :)

Nice, thanks! I had seen known bits had some handling of recurrences already, they didn't necessarily include selects.
I think we can expand this to phi+phi for the fpclass case too relatively easily, which will help in more complex cases.

I rebased over 577c7dd7cc4c5a9f62f9654cfa30ee9d55709426, which added a phase-ordering test.

arsenm commented:
As far as the regression on #83200 is concerned, I think it needs a local flag handling fix

arsenm reviewed (commented):
Needs a more targeted, dedicated test

davemgreen commented:
> As far as the regression on https://github.com/llvm/llvm-project/pull/83200 is concerned, I think it needs a local flag handling fix

Yeah there are probably a few ways to fix it, selects are often missing fmfs due to them coming from control flow.  This seemed generally useful.

> Needs a more targeted, dedicated test

I've tried to add some Attributor tests, which I see have been used to test similar patches in the past. Let me know if there is a better way.

goldsteinn commented on llvm/lib/Analysis/ValueTracking.cpp:
```diff
@@ -6003,6 +6003,13 @@ void computeKnownFPClass(const Value *V, const APInt &DemandedElts,
+        // operand to break the recursion.
+        Value *V;
+        if (match(IncValue, m_Select(m_Value(), m_Specific(P), m_Value(V))) ||
+            match(IncValue, m_Select(m_Value(), m_Value(V), m_Specific(P))))
```
Just a nit, but maybe this means we should add `m_c_Select` matcher.

arsenm replied:
It would be hard to represent that with the negated condition I think

goldsteinn replied:
Agreed, think it's probably not practical

davemgreen committed:
Add an extra commuted select test.
